enhancements/installer/storage-class-encrypted.md

```diff
@@ -0,0 +1,337 @@
+
+### Non-Goals
+* Other clouds besides AWS, GCP, & Azure are out of scope
+
+
+## Proposal
+
+### Workflow Description
+
+1. Cluster creator specifies (aws|gcp|azure) user-managed encryption keys 
+in `defaultMachinePlatform` of install config.
+2. Installer creates a manifest for `ClusterCSIDriver` which includes encryption keys in spec.
+3. Cluster-storage-operator/CSI drivers use installer-created `ClusterCSIDriver`
+to create default storage class using user-encrypted keys.
+
+
+### API Extensions
+
+This enhancement extends the [`ClusterCSIDriver` API object](https://github.com/openshift/api/blob/efeef9d83325a0ba12a636e5efd7d9414036f9f7/operator/v1/types_csi_cluster_driver.go).
+
+AWS, Azure & GCP will be added as new valid values for the `CSIDriverType` union discriminator
+in the CSIDriverConfigSpec:
+
+```go
+// CSIDriverType indicates type of CSI driver being configured.
+// +kubebuilder:validation:Enum=AWS;Azure;GCP;vSphere
+type CSIDriverType string
+
+const (
+  AWSDriverType     CSIDriverType = "AWS"
+  AzureDriverType   CSIDriverType = "Azure"
+  GCPDriverType     CSIDriverType = "GCP"
+  VSphereDriverType CSIDriverType = "vSphere" // existing platform
+)
+
+// CSIDriverConfigSpec defines configuration spec that can be
+// used to optionally configure a specific CSI Driver.
+// +union
+type CSIDriverConfigSpec struct {
+  // driverType indicates type of CSI driver for which the
+  // driverConfig is being applied to.
+  // Valid values are: AWS, Azure, GCP, vSphere.
+  // Consumers should treat unknown values as a NO-OP.
+  // +kubebuilder:validation:Required
+  // +unionDiscriminator
+  DriverType CSIDriverType `json:"driverType"`
+
+  // AWS is used to configure the AWS CSI driver.
+  // +optional
+  AWS *AWSCSIDriverConfigSpec `json:"aws,omitempty"`
+
+  // Azure is used to configure the Azure CSI driver.
+  // +optional
+  Azure *AzureCSIDriverConfigSpec `json:"azure,omitempty"`
+
+  // GCP is used to configure the GCP CSI driver.
+  // +optional
+  GCP *GCPCSIDriverConfigSpec `json:"gcp,omitempty"`
+
+  ... exisiting vsphere field ... 
+}
+```
+
+ and fields for the encryption keys will be added
+to platform-specific config specs:
+
+#### AWS
+
+```go
+// AWSCSIDriverConfigSpec defines properties that can be configured for the AWS CSI driver.
+type AWSCSIDriverConfigSpec struct {
+    // kmsKey sets the cluster default storage class to encrypt volumes with a user-defined KMS key,
+    // rather than the default KMS key used by AWS.
+    // +optional
+    KMSKey AWSKMSKey `json:"kmsKey,omitempty"`
+}
+```
+
+```go
+// AWSKMSKey contains the component parts that comprise an ARN for a KMS key.
+// According to https://docs.aws.amazon.com/service-authorization/latest/reference/list_awskeymanagementservice.html
+// the format for a KMS key ARN is:
+// arn:${Partition}:kms:${Region}:${Account}:key/${KeyId}
+// When constructing the ARN, the partition can be determined from the region.
```

Comment:
Breaking the ARN up into fields raises a couple of questions for me, although I really don't have a problem deferring this discussion to the openshift/api review.
1) How do we derive `Partition` from `Region`? Can the installer provide the `Partition` field too since it's presumably coming from something like `kmsKeyARN: arn:aws:kms:us-east-2:563456982459:key/4f5265b4-16f7-xxxx-xxxx-xxxxxxxxxxxx` in the install-config?
2) Will we support aliases as well, like `arn:${Partition}:kms:${Region}:${Account}:alias/${Alias}`? I think this is supported today (I have created a SC manually with such an ARN), but that does slightly complicate how we decompose/recompose this ARN string.

Reply:
Or maybe a more general question that I don't quite understand: what do we have to gain by decomposing and recomposing the ARN string?

Reply:
> How do we derive `Partition` from `Region`? 
Can use [`PartitionForRegion`](https://docs.aws.amazon.com/sdk-for-go/api/aws/endpoints/#PartitionForRegion) from the [endpoints package](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/pkg/cloud/cloud.go#L31)? Here is some [prior art in the installer](https://github.com/openshift/installer/blob/master/pkg/asset/installconfig/aws/validation.go#L119-L123). 

> Can the installer provide the `Partition` field too since it's presumably coming from something like `kmsKeyARN: arn:aws:kms:us-east-2:563456982459:key/4f5265b4-16f7-xxxx-xxxx-xxxxxxxxxxxx` in the install-config?
If we don't define a partition field in this API, the installer (or users) cannot pass it in.

> Or maybe a more general question that I don't quite understand: what do we have to gain by decomposing and recomposing the ARN string?
+1 to Joel's point, the perspective should be what is best UX for users when entering the values, which can be done either at install time or day 2. In this enhancement we are specifically defining the API that will be used day 2. Although we would be "decomposing" and "recomposing", we could change the install config API so it matches this API, if we think it's better for users. 

> Will they easily know the ARN and the correct format for it? Or are they more likely able to answer, region, accound and key id?

IMHO I think the following API for the AWS CSI driver is the most user friendly, and sufficient:
```
keyID - required
account - optional, defaults to account creds 
```
The keyID is the most easily accessible piece of information (it is available from the list view on the aws portal). We need account to offer cross-account keys, but the creds account is a sound default. I believe KMS keys are region restricted, so the region (and transitively the partition) can be inferred from the cluster region. So the values in the API would be much shorter than the less ergonomic, redundant ARNs.

This would be my preferred API and what I think is best for the users, but it is more work for the implementation on the CSI driver side. If you all agree, and think the work is not an obstacle, I think we should go this route unless you see any further issues. 

My opinion is that breaking up the ARN as it stands is slightly less user-friendly than allowing the complete ARN, but I think it gets us closer to the user friendly API described above and allows future work. It would be best to do it right the first time, but I know this enhancement has been taking a long time already...

Reply:
> Will we support aliases as well, like `arn:${Partition}:kms:${Region}:${Account}:alias/${Alias}`? I think this is supported today (I have created a SC manually with such an ARN), but that does slightly complicate how we decompose/recompose this ARN string.

It seems like the current API (proposed here) would need to be adjusted if we are accepting aliases. So the csi driver accepts Alias ARNs? The [csi driver parameter docs](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/master/docs/parameters.md) say:

> "kmsKeyId" |   |   | The full ARN of the key to use when encrypting the volume. If not  specified, AWS will use the default KMS key for the region the volume is  in. This will be an auto-generated key called /aws/ebs if not changed.
-- | -- | -- | --

Which is a little ambiguous to me. I wish they would say "the full ARN or Alias ARN", as these two things are distinguished in AWS docs, but that is probably just my being nitpicky. 

If we are accepting an alias, I think we have three options:

1. Discriminated union - user supplies alias xor key id (this is my preference)
2. we keep a single field (rename `keyID`) that will accept either alias or key id and we infer from the value format whether it is a key id or alias
3. back to accepting a full ARN as a string

Reply:
I think easiest to the user is to have a consistent API.
If MachineConfig uses [AWSResourceReference](https://github.com/openshift/api/blob/6ba31fa438f20d6c822ef47ad3d771309cb5216b/machine/v1beta1/types_awsprovider.go#L123), then its storage counterpart should use the same or something very similar.

Reply:
Not sure if it was mentioned before but currently - AWS CSI Operator does not talk to cloudprovider and does not even have aws-sdk vendored. We would prefer to keep it that way.

Reply:
> Not sure if it was mentioned before but currently - AWS CSI Operator does not talk to cloudprovider and does not even have aws-sdk vendored. We would prefer to keep it that way.

I meant to ask about that. I saw the sdk in the driver, but not the operator. So that would affect the ability to grab the partition and the account.

Reply:
IMO, the simplest and easiest to use option is to accept the full ARN as a string. As Jan points out, this is also consistent with how a user [provides the key to the installer](https://access.redhat.com/solutions/4827341) for the root volume today.

Looking at the key in the AWS console, the ARN is prominently displayed along with a 'copy' button to copy the ARN string to the clipboard. The Key ID and Alias are also prominently displayed, but the only way I see as a user to get the Partition, Region, and Account associated with the key I'm looking at in the console is to derive it from the ARN string. Parsing those fields out of the ARN and copying them individually seems like a worse user experience to me than to copy/paste the ARN string from AWS into the install-config.

Similar story for the CLI, `aws kms list-keys` shows two fields: `KeyId` and `KeyArn`.

Reply:
>  Parsing those fields out of the ARN and copying them individually seems like a worse user experience to me than to copy/paste the ARN string from AWS into the install-config.

+1 to this. I also think, all things considered, we should go with the ARN field. 

> but the only way I see as a user to get the Partition, Region, and Account associated with the key I'm looking at in the console is to derive it from the ARN string. 

The point I was trying to make is that *ideally* users shouldn't be required to provide those three fields at all, because it should already be known (except in the case when users are using cross-account KMS keys they would have to indeed parse that from the ARN). So in the case of keyID/alias vs ARN, I think keyID/alias is better. 

But this is too much of a stretch, in which case I think we should do the ARN.